File: bungee/src/main/java/net/impactdev/gts/bungee/listings/BungeeListing.java
package net.impactdev.gts.bungee.listings;

import net.impactdev.gts.api.listings.Listing;
import net.impactdev.gts.api.listings.entries.Entry;
import net.impactdev.gts.common.listings.JsonStoredEntry;
import net.impactdev.impactor.api.json.factory.JObject;

import java.time.LocalDateTime;
import java.util.UUID;

public abstract class BungeeListing implements Listing {

    private final UUID id;
    private final UUID lister;
    private final JsonStoredEntry entry;
    private final LocalDateTime published;
    private final LocalDateTime expiration;

    public BungeeListing(UUID id, UUID lister, JsonStoredEntry entry, LocalDateTime expiration) {
        this(id, lister, entry, LocalDateTime.now(), expiration);
    }

    public BungeeListing(UUID id, UUID lister, JsonStoredEntry entry, LocalDateTime published, LocalDateTime expiration) {
        this.id = id;
        this.lister = lister;
        this.entry = entry;
        this.published = published;
        this.expiration = expiration;
    }

    @Override
    public UUID getID() {
        return this.id;
    }

    @Override
    public UUID getLister() {
        return this.lister;
    }

    @Override
    public Entry<?, ?> getEntry() {
        return this.entry;
    }

    @Override
    public LocalDateTime getPublishTime() {
        return this.published;
    }

    @Override
    public LocalDateTime getExpiration() {
        return this.expiration;
    }

    @Override
    public int getVersion() {
        return 1;
    }

    @Override
    public JObject serialize() {
        JObject timings = new JObject()
                .add("published", this.getPublishTime().toString())
                .add("expiration", this.getExpiration().toString());

        return new JObject()
                .add("id", this.getID().toString())
                .add("lister", this.getLister().toString())
                .add("version", this.getVersion())
                .add("timings", timings)
                .add("entry", this.entry.getOrCreateElement());
    }
}

File: bungee/src/main/kotlin/net/impactdev/gts/bungee/listings/BungeeListing.kt
package net.impactdev.gts.bungee.listings

import net.impactdev.gts.api.listings.Listing
import net.impactdev.gts.api.listings.entries.Entry
import net.impactdev.gts.common.listings.JsonStoredEntry
import net.impactdev.impactor.api.json.factory.JObject
import java.time.LocalDateTime
import java.util.*

abstract class BungeeListing(
    override val iD: UUID?,
    override val lister: UUID?,
    private override val entry: JsonStoredEntry?,
    override val publishTime: LocalDateTime?,
    override val expiration: LocalDateTime?
) : Listing {

    constructor(id: UUID?, lister: UUID?, entry: JsonStoredEntry?, expiration: LocalDateTime?) : this(
        id,
        lister,
        entry,
        LocalDateTime.now(),
        expiration
    ) {
    }

    override fun getEntry(): Entry<*, *>? {
        return entry
    }

    override val version: Int
        get() = 1

    override fun serialize(): JObject? {
        val timings = JObject()
            .add("published", publishTime.toString())
            .add("expiration", expiration.toString())
        return JObject()
            .add("id", iD.toString())
            .add("lister", lister.toString())
            .add("version", version)
            .add("timings", timings)
            .add("entry", entry.getOrCreateElement())
    }
}